File: lib/features/shop/models/product.dart
class Product {
  const Product({
    required this.id,
    required this.name,
    required this.price,
    required this.stock,
    required this.isActive,
    this.description,
    this.imageUrl,
    this.categoryId,
  });

  final int id;
  final String name;
  final double price;
  final int stock;
  final bool isActive;
  final String? description;
  final String? imageUrl;
  final int? categoryId;

  factory Product.fromJson(Map<String, dynamic> json) => Product(
        id: _asInt(json['id']),
        name: json['name'] as String? ?? '',
        price: _asDouble(json['price']),
        stock: _asInt(json['stock'] ?? 0),
        isActive: (json['isActive'] ?? json['is_active']) == 1 ||
            json['isActive'] == true ||
            json['is_active'] == true,
        description: json['description'] as String?,
        imageUrl: json['image_url'] as String? ?? json['imageUrl'] as String?,
        categoryId: json['category_id'] != null
            ? _asInt(json['category_id'])
            : null,
      );
}

int _asInt(dynamic value) {
  if (value is int) return value;
  if (value is num) return value.toInt();
  if (value is String) {
    final parsed = int.tryParse(value);
    if (parsed != null) return parsed;
  }
  return 0;
}

double _asDouble(dynamic value) {
  if (value is double) return value;
  if (value is num) return value.toDouble();
  if (value is String) {
    final parsed = double.tryParse(value);
    if (parsed != null) return parsed;
  }
  return 0;
}

File: backend/src/entities/Product.ts
import {
  Column,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { ProductCategory } from './ProductCategory';

@Entity({ name: 'products' })
export class Product {
  @PrimaryGeneratedColumn({ type: 'bigint' })
  id!: number;

  @Column({ name: 'category_id', type: 'bigint', nullable: true })
  categoryId!: number | null;

  @ManyToOne(() => ProductCategory, { onDelete: 'SET NULL' })
  @JoinColumn({ name: 'category_id' })
  category!: ProductCategory | null;

  @Column({ length: 180 })
  name!: string;

  @Column({ type: 'text', nullable: true })
  description!: string | null;

  @Column({ type: 'decimal', precision: 10, scale: 2 })
  price!: string;

  @Column({ type: 'int', default: 0 })
  stock!: number;

  @Column({ name: 'image_url', type: 'varchar', length: 255, nullable: true })
  imageUrl!: string | null;

  @Column({ name: 'is_active', type: 'tinyint', width: 1, default: 1 })
  isActive!: boolean;

  @Column({ name: 'created_at', type: 'datetime', nullable: true })
  createdAt!: Date | null;

  @Column({ name: 'updated_at', type: 'datetime', nullable: true })
  updatedAt!: Date | null;
}
